### PR TITLE
Feature/History panel with older requests

### PR DESCRIPTION
**Pull request by epilgrim**
panel looks like
image
and if you click the UUID, you see previous requests data:
image
i'm not sure about the "hint" though... do you have a better idea how to highlight that we are not in the current request?
also, I can make a click anywhere in the row instead of in the uuid.

---
This change is [Reviewable](https://reviewable.io/reviews/kagux/ex_debug_toolbar/54)

### CODE REVIEW

**Comment by kagux** [comment]
addresses #48

**Comment by kagux** [comment]
Hey,

This is very cool! I'll go through code later today or tomorrow. A few things I want to mention now:
* uuid is probably not a valuable info for users and I'd just hide this column
* on the other hand, things that could be useful to see: response code, time
* i like current hint, maybe try to play with text color( text warning?) and weight,but nothing more. 
* it'd be super convenient to have a shortcut next to hint that takes you back to current request.
* we need to pick different icons for history and timing, I know it's challenging
* how come path is empty for many requests?

**Comment by epilgrim** [comment]
hey, regarding:
empty path: i'm not sure why, but sometimes when navigating back to the root, the `/` appears, and sometimes it is empty. haven't delved into it, but empty is the same as `/`
current hint taking you to current request. I removed it, because i was doing a hack to insert the uuid of the original request each time we got html response from the server. but I just thought of a better way :) so it won't look like a hack. I'll do it later
different icons: yeah... i was procrastinating :D

**Comment by epilgrim** [comment]
what about now? new hint pointing to the original, new glyphicon and displayed data
image

**Comment by epilgrim** [comment]
mmm, the empty path does not only happens when pointing to root url. In the last picture, that empty should have been '/ecto'... but that data is not in the connection stored in the toolbar

**Comment by kagux** [comment]
Nice! 
* I think "Back to Current Request" should be more prominent (bold or warning color) or maybe have a link or button style. What do you think?
* Response code should be colored like it is in the toolbar
* I'd reorder columns to be [path, method, status, duration, timestamp]

I have no idea why would conn not have a proper path :\ sounds like a bug somewhere

**Comment by kagux** [comment]
Reviewed 3 of 9 files at r1, 6 of 6 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[test/views/toolbar_view_test.exs, line 115 at r2](https://reviewable.io:443/reviews/kagux/ex_debug_toolbar/54#-KtfQPnPtKWv7zA1dJc5:-KtfQPnPtKWv7zA1dJc6:b-whcgov) ([raw file](https://github.com/kagux/ex_debug_toolbar/blob/de328b582cc1d498e4f402f358f2964feb662dba/test/views/toolbar_view_test.exs#L115)):*
> ```Elixir
>     assigns = [
>       request: request,
>       history: [request],
> ```

will this actually test history panel? maybe we need another request to make sure it renders all panel elements

---

*[web/static/js/toolbar.js, line 19 at r2](https://reviewable.io:443/reviews/kagux/ex_debug_toolbar/54#-KtfQi9_Y0eVJrDL0e8h:-KtfQi9ay8tYoZfc4S_j:bd5wqwb) ([raw file](https://github.com/kagux/ex_debug_toolbar/blob/de328b582cc1d498e4f402f358f2964feb662dba/web/static/js/toolbar.js#L19)):*
> ```JavaScript
>     this.toolbar = $("<div>", {id: "ex-debug-toolbar"});
>     $("body").append(this.toolbar);
>     this._setupHistoryListeners(this.toolbar);
> ```

other private methods are not underscored, don't think it's worth it.

---

*[web/static/js/toolbar/breakpoints_panel.js, line 7 at r2](https://reviewable.io:443/reviews/kagux/ex_debug_toolbar/54#-KtfS8K6GAMZixs5ccim:-KtfS8K6GAMZixs5ccin:b-7m74ox) ([raw file](https://github.com/kagux/ex_debug_toolbar/blob/de328b582cc1d498e4f402f358f2964feb662dba/web/static/js/toolbar/breakpoints_panel.js#L7)):*
> ```JavaScript
> 
> class BreakpointsPanel {
>   constructor(socket, toolbar) {
> ```

nice!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/kagux/ex_debug_toolbar/54)*

**Comment by epilgrim** [comment]
I tried with button, but I don't like how it looks. Also, this part of the toolbar should be minimalistic.... simple text should be fine.
Regarding colors, I like the black, but this others look good as well:

image
image
image

**Comment by kagux** [comment]
those are not bootstrap colors, are they? I meant one of those "text-warning" classes.
what about "btn btn-link" class?

**Comment by epilgrim** [comment]
If It is btn-link, I still have to put the color manually, as the link does not accept the warning class.
as text-warning, it looks 
image
which doesn't look good :D

**Comment by epilgrim** [comment]
regarding the color of the response, currently it is always green:
image
should I create a view helper coloring this based on the response code in this PR?

**Comment by kagux** [comment]
yeah, it doesn't :)
with btn link I meant the default link color
anyways, I don't want to block you over this minor thing. let's keep the one you prefer!

File: web/static/js/toolbar.js [hunk 1]
```diff
@@ -11,13 +11,20 @@ import 'prismjs/plugins/line-highlight/prism-line-highlight';
 class App {
   constructor(opts) {
     this.opts = opts;
+    this.originalRequestId = opts.requestId;
     this.resetActivePanel();
+    this.socket = this._initSocket();
```
Comment (kagux) [inline web/static/js/toolbar.js]: looks like a leftover

**Comment by kagux** [comment]
I thought view already had a helper for status :) I probably forgot to add it, my bad! So yeah, please, add it.

**Comment by epilgrim** [comment]
Review status: 4 of 11 files reviewed at latest revision, 4 unresolved discussions.

---

*[test/views/toolbar_view_test.exs, line 115 at r2](https://reviewable.io:443/reviews/kagux/ex_debug_toolbar/54#-KtfQPnPtKWv7zA1dJc5:-Ktjpl_ZP68pjCOfVVej:b-1leexe) ([raw file](https://github.com/kagux/ex_debug_toolbar/blob/de328b582cc1d498e4f402f358f2964feb662dba/test/views/toolbar_view_test.exs#L115)):*
<details><summary><i>Previously, kagux (Boris Mikhaylov) wrote…</i></summary><blockquote>

will this actually test history panel? maybe we need another request to make sure it renders all panel elements
</blockquote></details>

it tests the history panel the same way as other tests. it checks that it renders. There is nothing special here that would change if you have 1 or 2 requests in the history

---

*[web/static/js/toolbar.js, line 19 at r2](https://reviewable.io:443/reviews/kagux/ex_debug_toolbar/54#-KtfQi9_Y0eVJrDL0e8h:-Ktjpw-IYmMiRpXcBdYb:b-896fix) ([raw file](https://github.com/kagux/ex_debug_toolbar/blob/de328b582cc1d498e4f402f358f2964feb662dba/web/static/js/toolbar.js#L19)):*
<details><summary><i>Previously, kagux (Boris Mikhaylov) wrote…</i></summary><blockquote>

other private methods are not underscored, don't think it's worth it.
</blockquote></details>

Done.

---

*[web/static/js/toolbar.js, line 16 at r3](https://reviewable.io:443/reviews/kagux/ex_debug_toolbar/54#-KtbmVNXPMgNoB33zRq1-r3-16:-Ktjpv789i2FYTmDBCY8:b-896fix) ([raw file](https://github.com/kagux/ex_debug_toolbar/blob/9a99a0e950c63497c3ba1c63b3024d970054020b/web/static/js/toolbar.js#L16)):*
<details><summary><i>Previously, kagux (Boris Mikhaylov) wrote…</i></summary><blockquote>

looks like a leftover 
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/kagux/ex_debug_toolbar/54)*

File: web/views/toolbar_view.ex [hunk 2]
```diff
@@ -30,6 +30,17 @@ defmodule ExDebugToolbar.ToolbarView do
...
   end
 
+  def conn_status_color_class(%{status: status}) do
+    cond do
+      status <= 199 -> "label-info"
+      status <= 299 -> "label-success"
+      status <= 399 -> "label-info"
+      status <= 499 -> "label-danger"
+      status <= 599 -> "label-danger"
+      true -> "label-danger"
```
Comment (kagux) [inline web/views/toolbar_view.ex]: why not merge last 3 ones?

File: test/views/toolbar_view_test.exs [hunk 3]
```diff
@@ -7,111 +7,123 @@ defmodule ExDebugToolbar.ToolbarViewTest do
...
+        pid: self(),
+        line: 5,
+        file: "test.ex",
+        code_snippet: [{"a = [1, 2]", 5}],
+        env: __ENV__,
+        binding: binding(),
+        inserted_at: NaiveDateTime.utc_now
+      }
+      assert %Request{} |> render(breakpoints: [breakpoint]) |> is_bitstring
+    end
```
Comment (kagux) [inline test/views/toolbar_view_test.exs]: I think we need a similar test that renders with multiple requests in history to ensure panel renders without errors

File: web/templates/toolbar/show/_history.html.eex [hunk 4]
```diff
@@ -0,0 +1,34 @@
+<span class="toolbar-item" data-toggle="panel" data-title="All Requests">
+  <i class="glyphicon glyphicon-list-alt" aria-hidden="true"></i>
+  <%= length @history %>
+</span>
+<div class="panel panel-default navbar-fixed-bottom">
+  <div class="panel-heading">
+    <h3 class="panel-title">All Requests</h3>
```
Comment (kagux) [inline web/templates/toolbar/show/_history.html.eex]: seeing as we call it history should we name the panel history as well?

**Comment by epilgrim** [comment]
Colored the whole row when status is not success (all green didn't look good)
Also added controller name. 
Putting the method before path was on purpose, i do think it reads better
image

**Pull request merged**